sweep/aggregator.go

```diff
@@ -0,0 +1,349 @@
+	// is done by determining the fee rate bucket they should belong in.
+	for op, input := range inputs {
+		feeRate, err := input.params.Fee.Estimate(
+			s.FeeEstimator, s.MaxFeeRate,
+		)
+		if err != nil {
+			log.Warnf("Skipping input %v: %v", op, err)
+			continue
+		}
+
+		// Only try to sweep inputs with an unconfirmed parent if the
+		// current sweep fee rate exceeds the parent tx fee rate. This
+		// assumes that such inputs are offered to the sweeper solely
+		// for the purpose of anchoring down the parent tx using cpfp.
+		parentTx := input.UnconfParent()
+		if parentTx != nil {
+			parentFeeRate :=
+				chainfee.SatPerKWeight(parentTx.Fee*1000) /
+					chainfee.SatPerKWeight(parentTx.Weight)
+
+			if parentFeeRate >= feeRate {
+				log.Debugf("Skipping cpfp input %v: "+
+					"fee_rate=%v, parent_fee_rate=%v", op,
+					feeRate, parentFeeRate)
+
+				continue
+			}
+		}
+
+		feeGroup := s.bucketForFeeRate(feeRate)
+
+		// Create a bucket list for this fee rate if there isn't one
+		// yet.
+		buckets, ok := bucketInputs[feeGroup]
+		if !ok {
+			buckets = &bucketList{}
+			bucketInputs[feeGroup] = buckets
+		}
+
+		// Request the bucket list to add this input. The bucket list
+		// will take into account exclusive group constraints.
+		buckets.add(input)
+
+		input.lastFeeRate = feeRate
+		inputFeeRates[op] = feeRate
+	}
+
+	// We'll then determine the sweep fee rate for each set of inputs by
+	// calculating the average fee rate of the inputs within each set.
+	inputClusters := make([]inputCluster, 0, len(bucketInputs))
+	for _, buckets := range bucketInputs {
+		for _, inputs := range buckets.buckets {
+			var sweepFeeRate chainfee.SatPerKWeight
+			for op := range inputs {
+				sweepFeeRate += inputFeeRates[op]
+			}
+			sweepFeeRate /= chainfee.SatPerKWeight(len(inputs))
+			inputClusters = append(inputClusters, inputCluster{
+				sweepFeeRate: sweepFeeRate,
+				inputs:       inputs,
+			})
+		}
+	}
+
+	return inputClusters
+}
+
+// bucketForFeeReate determines the proper bucket for a fee rate. This is done
+// in order to batch inputs with similar fee rates together.
+func (s *SimpleAggregator) bucketForFeeRate(
+	feeRate chainfee.SatPerKWeight) int {
+
+	relayFeeRate := s.FeeEstimator.RelayFeePerKW()
+
+	// Create an isolated bucket for sweeps at the minimum fee rate. This
+	// is to prevent very small outputs (anchors) from becoming
+	// uneconomical if their fee rate would be averaged with higher fee
+	// rate inputs in a regular bucket.
+	if feeRate == relayFeeRate {
+		return 0
+	}
+
+	return 1 + int(feeRate-relayFeeRate)/s.FeeRateBucketSize
+}
+
+// mergeClusters attempts to merge cluster a and b if they are compatible. The
+// new cluster will have the locktime set if a or b had a locktime set, and a
+// sweep fee rate that is the maximum of a and b's. If the two clusters are not
+// compatible, they will be returned unchanged.
+func mergeClusters(a, b inputCluster) []inputCluster {
```

Comment:
since this function is assessing compatibility based off of locktime equality, shouldn't it also assess feerate compatibility using the cluster size as an argument?

Reply:
nvm we aren't changing logic in this commit

Reply:
yeh this is sorta a port of the old logic to the new interface